consensus/src/detail/tests/double_spend_tests.rs

```diff
@@ -0,0 +1,314 @@
+#[allow(clippy::eq_op)]
+fn double_spend_tx_in_the_same_block() {
+    common::concurrency::model(|| {
+        // Check is it correctly spend when a couple of transactions pointing on one output
+        // +--Block----------------+
+        // |                       |
+        // | +-------tx-1--------+ |
+        // | |input = prev_block | |
+        // | +-------------------+ |
+        // |                       |
+        // | +-------tx-2--------+ |
+        // | |input = tx1        | |
+        // | +-------------------+ |
+        // |                       |
+        // | +-------tx-3--------+ |
+        // | |input = tx1        | |
+        // | +-------------------+ |
+        // +-----------------------+
+
+        let mut consensus = setup_consensus();
+        let receiver = random_address(&consensus.chain_config);
+
+        let prev_block_tx_id =
+            consensus.chain_config.genesis_block().transactions().get(0).unwrap().get_id();
+
+        // Create first tx
+        let first_tx = Transaction::new(
+            0,
+            vec![TxInput::new(
+                OutPointSourceId::Transaction(prev_block_tx_id),
+                0,
+                random_witness(),
+            )],
+            vec![TxOutput::new(Amount::from_atoms(12345678912345), receiver.clone())],
+            0,
+        )
+        .expect(ERR_CREATE_TX_FAIL);
+        let first_tx_id = first_tx.get_id();
+
+        // Create second tx
+        let second_tx = Transaction::new(
+            0,
+            vec![TxInput::new(first_tx_id.clone().into(), 0, vec![])],
+            vec![TxOutput::new(Amount::from_atoms(987654321), receiver.clone())],
+            0,
+        )
+        .expect(ERR_CREATE_TX_FAIL);
+
+        // Create third tx
+        let third_tx = Transaction::new(
+            123456789,
+            vec![TxInput::new(first_tx_id.into(), 0, vec![])],
+            vec![TxOutput::new(Amount::from_atoms(987654321), receiver)],
+            0,
+        )
+        .expect(ERR_CREATE_TX_FAIL);
+
+        // Create tx that pointing to the previous tx
+        let block = Block::new(
+            vec![first_tx, second_tx, third_tx],
+            Some(Id::new(
+                &consensus.chain_config.genesis_block().get_id().get(),
+            )),
+            time::get() as u32,
+            ConsensusData::None,
+        )
+        .expect(ERR_CREATE_BLOCK_FAIL);
+        assert!(consensus.process_block(block, BlockSource::Local).is_err());
+        assert_eq!(
+            consensus
+                .blockchain_storage
+                .get_best_block_id()
+                .expect(ERR_BEST_BLOCK_NOT_FOUND)
+                .expect(ERR_STORAGE_FAIL),
+            consensus.chain_config.genesis_block().get_id()
+        );
+    });
+}
+
+#[test]
+fn double_spend_tx_in_another_block() {
+    common::concurrency::model(|| {
+        // Check is it correctly spend when a couple of transactions in a different blocks pointing on one output
```

Comment:
I'm not sure I understand this comment

Reply:
I would rephrase: A few different blocks use an exactly one output. Check whether it's might spend

Reply:
Is it works for you? :)

Reply:
Fixed